File: python基础/数据结构与算法/排序/quick_sort.py
# 按基准值，左大右小摆放，分而治之
def quick_sort(arr):
    if len(arr)>1: #递归入口
        mid = arr.pop() #基值
        l, r = [], []
        for i in arr:
            if i>mid:
                r.append(i) #大于基准值放右边
            else:
                l.append(i) #小于基准值放左边
        return quick_sort(l) + [mid] + quick_sort(r)
    else:
        return arr

print(quick_sort([1,3,1,3,2,4]))

File: Golang基础/算法/排序/quickSort.go
// 分治(nlog2 n; n^2; nlog2 n; log2 n)

package main

import (
	"fmt"
	_ "math/rand"
)

func QuickSort(s []int) []int {
	if length := len(s); length > 1 {
		var larr, rarr, mid []int
		n := 0
		// n := rand.Int() % length
		splitdata := s[n]
		mid = append(mid, splitdata)

		for i, v := range s {
			if i == n {
				continue
			}

			if v > splitdata {
				rarr = append(rarr, v)
			} else if v < splitdata {
				larr = append(larr, v)
			} else {
				mid = append(mid, v)
			}
		}
		return append(append(QuickSort(larr), mid...), QuickSort(rarr)...)
	} else {
		return s
	}
}

func main() {
	arr := []int{1, 4, 54, 2, 7, 3, 2, 6}
	fmt.Println(QuickSort(arr))
}
